_drafts/best-of-binutils.md

```diff
@@ -0,0 +1,1011 @@
+    <11>   DW_AT_name        : (indirect string, offset: 0x4576): main.c
+    <15>   DW_AT_comp_dir    : (indirect string, offset: 0x3692): example/best-of-binutils
+    <19>   DW_AT_ranges      : 0x18
+    <1d>   DW_AT_low_pc      : 0x0
+    <21>   DW_AT_stmt_list   : 0x0
+    <25>   DW_AT_GNU_macros  : 0x0
+[...]
+```
+
+#### Auditing Compiler Flag Settings
+
+There's certain DIEs that are often interesting to look at in isolation. For example, you can
+look at `DW_AT_producer` DIE on GCC and Clang to see the compiler flags which were used.
+
+```bash
+$ arm-none-eabi-readelf --debug-dump nrf52_example/build/nrf52.elf |
+   grep DW_AT_producer.*:
+
+    <c>   DW_AT_producer    :
+    (indirect string, offset: 0x2cc5): GNU C17 9.2.1 20191025 (release)
+    [ARM/arm-9-branch revision 277599]
+    -mcpu=cortex-m4 -mfloat-abi=hard -mfpu=fpv4-sp-d16 -mthumb
+    -march=armv7e-m+fp -g3 -Os -ffunction-sections -fdata-sections
+```
+
+> NOTE: For more details about good compiler flag settings for embedded devices in general, check out our
+> post on the topic [here]({% post_url 2019-10-22-best-and-worst-gcc-clang-compiler-flags %}).
+
+#### Read GNU Build ID
+
+When using a GNU build ID, the `-n` option can be used to quickly
+print note sections and see the value
+
+```bash
+arm-none-eabi-readelf -n nrf52_example/build/nrf52.elf
+
+Displaying notes found in: .gnu_build_id
+  Owner                Data size    Description
+  GNU                  0x00000014	NT_GNU_BUILD_ID (unique build ID bitstring)
+  Build ID: c61acc0ab0602a9f53bcc36b91d017834129fea1
+```
+
+> NOTE: To learn more about GNU Build IDs and how to easily add one to your project, check out
+> [this article]({% post_url 2019-05-29-gnu-build-id-for-firmware %}).
+
+### strings
+
+As the name implies, `strings` will display all the strings in a binary. We can use the tool to:
+
+1. To check for security holes in a binary (i.e encryption keys, hard coded passwords, etc)
+2. As a mechanism for quickly saving [code space]({% post_url
+   2019-08-20-code-size-optimization-gcc-flags %}). More often than not there are a couple
+   unexpectedly long file names or command line strings in a binary that can be trimmed.
+
+There's two particularly helpful arguments:
+
+- `-d` - Causes only data sections (code in the actual binary) to be scanned instead of the entire
+  info rather than parts of the ELF that are not in the binary.
+- `-n <number>` - Can be used to control the minimum string length scanned for.
+
+#### Leveraging strings to Save Code
+
+Let's take a look at the longest strings in the Zephyr `samples/net/wifi` application `west build -b disco_l475_iot1`:
+
+```bash
+$ cd <zephyr_project>/samples/net/wifi
+$ west build -b disco_l475_iot1
+
+$ arm-none-eabi-strings -n5 -d build/zephyr/zephyr.elf |
+   awk '{ print length, $0 }' | sort -n
+// ...
+94 Ctrl+a, Ctrl+b, Ctrl+c, Ctrl+d, Ctrl+e, Ctrl+f, Ctrl+k, Ctrl+l, Ctrl+n, Ctrl+p, Ctrl+u, Ctrl+w
```

Comment:
What? No reverse sort?!

Reply:
hehe, my thought was then you would probably want a less pipe so you didn't have to scroll up